File: yaml_parsers/ruby/parser.rb
require 'yaml'
require 'timeout'


def parse_yaml(input_file, output_file)
    data = nil # Defining the variable in the outer scope
    begin
      # Set a timeout of 5 seconds
      Timeout.timeout(120) do
        data = YAML.load_file(input_file)
      end
    rescue Timeout::Error
      puts "Task timed out!"
      return # Exit the method early if there was a timeout
    end

    if data
    File.open(output_file, 'w') { |file| file.write(data.to_yaml) }
  else
    puts "Failed to load data."
  end
end

if __FILE__ == $0
    input_file = ARGV[0]
    output_file = ARGV[1]
    if input_file.nil? || output_file.nil?
        puts "Please provide input and output file paths."
        exit 1
    end
    parse_yaml(input_file, output_file)
end
    

File: yaml_parsers/python/parser.py
import sys
import yaml
import signal
    
# Function to handle the timeout
def handler(signum, frame):
    raise TimeoutError("Operation timed out after 2 minutes")



def parse_yaml(input_file, output_file):
    # Set the timeout handler to trigger after 120 seconds
    signal.signal(signal.SIGALRM, handler)
    signal.alarm(120)  # Set the alarm for 120 seconds (2 minutes)
    try:
        with open(input_file, 'r') as infile:
            data = yaml.safe_load(infile)
        with open(output_file, 'w') as outfile:
            yaml.dump(data, outfile)
    except TimeoutError as e:
        print(e)
    finally:
        # Disable the alarm
        signal.alarm(0)

if __name__ == "__main__":
    if (len(sys.argv) == 3):
        input_file = sys.argv[1]
        output_file = sys.argv[2]
        if input_file and output_file:
            parse_yaml(input_file, output_file)
    else:
        parse_yaml("/Users/a/PycharmProjects/yamlparsers/input.yaml", "output.yaml")
    
    